pkg/build/builder/cmd/scmauth/passwd.go

```diff
@@ -0,0 +1,75 @@
+package scmauth
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	UsernamePasswordName = "passwd"
+	UserPassGitConfig    = `# username and password git config
+[url "%[1]s"]
+   insteadOf = %[2]s
+`
+)
+
+// SSHPrivateKey implements SCMAuth interface for using SSH private keys.
+type UsernamePassword struct {
+	SourceURL url.URL
+}
+
+// Setup creates a wrapper script for SSH command to be able to use the provided
+// SSH key while accessing private repository.
+func (u UsernamePassword) Setup(baseDir string) error {
+	// Only apply to https and http URLs
+	if scheme := strings.ToLower(u.SourceURL.Scheme); scheme != "http" && scheme != "https" {
+		return nil
+	}
+	uname, password, err := readUsernamePassword(filepath.Join(baseDir, UsernamePasswordName))
+	if err != nil {
+		return err
+	}
+	gitconfig, err := ioutil.TempFile("", "gitpasswd")
+	if err != nil {
+		return err
+	}
+	defer gitconfig.Close()
+
+	usernamePasswordURL := u.SourceURL
+	usernamePasswordURL.User = url.UserPassword(uname, password)
+	gitconfig.WriteString(fmt.Sprintf(UserPassGitConfig, usernamePasswordURL.String(), u.SourceURL.String()))
+
+	return ensureGitConfig(gitconfig.Name())
+}
+
+// Name returns the name of this auth method.
+func (_ UsernamePassword) Name() string {
+	return UsernamePasswordName
+}
+
+func readUsernamePassword(file string) (username, password string, err error) {
```

Comment:
why are we making them pass in this format? is this something they are likely to have already?